### PR TITLE
Added shell executor class

### PR DESCRIPTION
**Pull request by tomaszrybicki**
Signed-off-by: Tomasz Rybicki <tomasz.rybicki@intel.com>

### CODE REVIEW

File: source/octf/utils/ShellExecutor.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	virtual ~ShellExecutor() = default;
+
+	/**
+	 * @brief Executes a command in shell by using popen()
+	 *
+	 * This effectively creates a pipe, forks and invokes the default shell
+	 *
+	 * @note Command and arguments are simply concatenated with spaces
+	 */
+	std::string executeCommand();
```
Comment (mariuszbarczak) [inline source/octf/utils/ShellExecutor.h]: I think just execute is enough

File: source/octf/utils/ShellExecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+	int read;
+	while ((read = fread(buffer.data(), 1, buffer.size(), pipe.get()))
+			== static_cast<int>(buffer.size())) {
+		result += buffer.data();
+	}
+
+	// Error occurred
+	if (ferror(pipe.get())) {
+		throw Exception("Could not read from started casadm process");
```
Comment (mariuszbarczak) [inline source/octf/utils/ShellExecutor.cpp]: please update comment, and maybe would be good print process name

File: source/octf/utils/ShellExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
+	if (!pipe) {
+		throw Exception("Call to popen() failed!");
+	}
+
+	int read;
+	while ((read = fread(buffer.data(), 1, buffer.size(), pipe.get()))
+			== static_cast<int>(buffer.size())) {
+		result += buffer.data();
```
Comment (mariuszbarczak) [inline source/octf/utils/ShellExecutor.cpp]: I'd use string::append method, data can be binary

File: source/octf/utils/ShellExecutor.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	// Error occurred
+	if (ferror(pipe.get())) {
+		throw Exception("Could not read from started casadm process");
+	}
+
+	// Not entire buffer was read at the end
+	if (feof(pipe.get()) && read > 0) {
+		buffer[read] = '\0';
+		result += buffer.data();
```
Comment (mariuszbarczak) [inline source/octf/utils/ShellExecutor.cpp]: I'd use string::append method, data can be binary

File: source/octf/utils/ShellExecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	std::string result;
+
+	std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
+	if (!pipe) {
+		throw Exception("Call to popen() failed!");
+	}
+
+	int read;
+	while ((read = fread(buffer.data(), 1, buffer.size(), pipe.get()))
+			== static_cast<int>(buffer.size())) {
```
Comment (mariuszbarczak) [inline source/octf/utils/ShellExecutor.cpp]: Do you need cast here?

File: source/octf/utils/ShellExecutor.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	command += m_args;
+
+	const char* cmd = command.c_str();
+
+	std::array<char, 128> buffer;
+	std::string result;
+
+	std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
+	if (!pipe) {
+		throw Exception("Call to popen() failed!");
```
Comment (mariuszbarczak) [inline source/octf/utils/ShellExecutor.cpp]: Maybe we can have dedicated exception?

File: source/octf/utils/ShellExecutor.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+	const std::string& getArgs() const;
```
Comment (mariuszbarczak) [inline source/octf/utils/ShellExecutor.h]: Don't we want to have inline getters and setters?

File: source/octf/utils/ShellExecutor.cpp (hunk 3 above)
Comment (tomaszrybicki) [reply source/octf/utils/ShellExecutor.cpp]: Good catch

File: source/octf/utils/ShellExecutor.cpp (hunk 5 above)
Comment (tomaszrybicki) [reply source/octf/utils/ShellExecutor.cpp]: Without a cast there is an unsigned with signed comparison. It does not matter now with current values but the buffer size may be changed in the future, and also the compiler warnings are suppressed.

**Pull request merged**